aws/resource_aws_iot_policy_test.go

```diff
@@ -0,0 +1,109 @@
+package aws
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/aws/aws-sdk-go/service/iot"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/terraform"
+	"regexp"
+)
+
+func TestAccAWSIoTPolicy_basic(t *testing.T) {
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSIoTPolicyDestroy_basic,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: testAccAWSIoTPolicy_basic,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSIoTPolicyExists_basic("aws_iot_policy.pubsub"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccAWSIoTPolicy_invalidJson(t *testing.T) {
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSIoTPolicyDestroy_basic,
+		Steps: []resource.TestStep{
+			{
+				Config:      testAccAWSIoTPolicyInvalidJsonConfig,
+				ExpectError: regexp.MustCompile("MalformedPolicyException.*"),
+			},
+		},
+	})
+}
+
+func testAccCheckAWSIoTPolicyDestroy_basic(s *terraform.State) error {
+	conn := testAccProvider.Meta().(*AWSClient).iotconn
+
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "aws_iot_policy" {
+			continue
+		}
+
+		out, err := conn.ListPolicies(&iot.ListPoliciesInput{})
+
+		if err != nil {
+			return err
+		}
+
+		for _, t := range out.Policies {
+			if *t.PolicyName == rs.Primary.ID {
+				return fmt.Errorf("IoT policy still exists:\n%s", t)
+			}
+		}
+
+	}
+
+	return nil
+}
+
+func testAccCheckAWSIoTPolicyExists_basic(name string) resource.TestCheckFunc {
```

Comment:
We usually leverage helpers like `resource.TestCheckResourceAttr` and `resource.TestCheckResourceAttrSet` to check whether the resource was recorded in the state.

The point of having a function like this is to allow more thorough check of the API/SDK struct, take a look how we do it in a VPC test: https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_vpc_test.go#L25-L26

It's not mandatory to perform such thorough checking, but if you decide not to do it there's no point of having this function when the standard helpers can do the job.